### PR TITLE
[GSoC] Setting up database classes and UI changes of the Custom Image Selector

### LINKED ISSUE #4764: GSoC 2022: Improve the custom picture selector
Students interested in GSoC 2022, welcome! Please read [the instructions to apply](https://github.com/commons-app/commons-app-documentation/blob/master/android/Students.md).
The set of enhancements below constitutes the proposed task that the chosen student will implement during GSoC 2022.

---

The custom picture selector appears when the user taps the `+` then taps the third button (at the bottom, a gallery icon with a small Commons icon overlay):

adb

Below is a list of enhancements that would make it more convenient for users:

# UI when within a folder

- At the bottom of the picker, show two buttons:
  - "Upload" to start the Upload Wizard with the selected pictures (replacing the already implemented ✔️ button currently available at the top-right).
  - "Mark as not for upload" to mark these pictures with a 🚫 overlay icon (and grey them out). For instance family pictures, not meant to be uploaded to Commons. This state must be remembered even after restarting the app. Tapping "Not for upload" when all selected pictures are already marked as "Not for upload" has the opposite effect, a bit like pressing "Italic" twice in a text editor makes the selected text in Italic then normal.
- Button to hide/unhide pictures that are either marked as "Not for upload" or have been uploaded already. Hiding them allows the contributor to more easily spot pictures they might want to upload.
  - For a good user experience, the screen should try to show the same point in time after changing this setting. For instance in the mockup below, after disabling this setting I should still see the pictures with no lower-right overlay, even if many pictures were removed above them. This may require adjusting the scroll position.
- Currently, when selecting 3 pictures, they are numbered "1", "2", "3". Instead, please make all numbers always show the number of selected pictures, for instance if 3 pictures are selected, they should all have the number "3", and be uploaded in the same order as their thumbnails are shown.
- Add a fast scrollbar (ideally bubble scroll showing the date at which the pictures shown were taken).

gsoc2022pictures

# UI when in full-screen mode

(fullscreen starts when you long-press any picture in the custom picker)

- Show a small overlay indicating whether the picture is selected or not, showing the number of selected pictures.
- Gestures:
  - Left/right swipe to go from one picture to another.
  - Swipe up to select the picture (the equivalent of tapping it in non-fullscreen mode) and show the next picture (the picture that follows in the folder view, skipping pictures that have either already been uploaded or ignored).
  - Swipe down to immediately mark that picture as "Not for upload" (the equivalent of selecting it then tapping "Mark as not for upload" in non-fullscreen mode), and show the next picture.

adb

### PR DESCRIPTION
**Pull request by Ayan-10**
Tasks

- [x] Create of all required database classes
- [x] Remove ✔️ button from toolbar
- [x] Add two buttons in the bottom layout for uploading and marking as not for upload

### CODE REVIEW

File: app/src/main/res/layout/activity_custom_selector.xml [hunk 1]
```diff
@@ -14,6 +14,52 @@
       android:layout_width="match_parent"
       android:layout_height="0dp"
       app:layout_constraintBottom_toBottomOf="parent"
+      app:layout_constraintBottom_toTopOf="@id/bottom_layout"
       app:layout_constraintTop_toBottomOf="@+id/toolbar_layout"/>
 
+    <androidx.constraintlayout.widget.ConstraintLayout
```
Comment (4D17Y4) [inline app/src/main/res/layout/activity_custom_selector.xml:20]: I would prefer the bottom layout to be in a separate file and include it here same as it is done with the toolbar. 
Makes the XML more modular and easy to read. 

@nicolas-raoul views?

**Review by 4D17Y4 — CHANGES_REQUESTED** [review]
Minor changes.

File: app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/CustomSelectorActivity.kt [hunk 2]
```diff
@@ -128,7 +129,7 @@ class CustomSelectorActivity: BaseActivity(), FolderClickListener, ImageSelectLi
         val back : ImageButton = findViewById(R.id.back)
         back.setOnClickListener { onBackPressed() }
 
-        val done : ImageButton = findViewById(R.id.done)
```
[the remark this answers is not in the document]
Comment (Ayan-10) [reply app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/CustomSelectorActivity.kt:131]: I didn't remove it. The button was hidden. Should I remove it?

File: app/src/main/res/layout/activity_custom_selector.xml [hunk 3]
```diff
@@ -14,6 +14,52 @@
...
+      app:layout_constraintStart_toStartOf="parent">
+
+        <Button
+          android:id="@+id/not_for_upload"
+          android:layout_width="0dp"
+          android:layout_height="wrap_content"
+          android:text="Mark as not for upload"
+          android:textColor="@color/white"
+          android:visibility="visible"
+          android:layout_margin="5dp"
```
[the remark this answers is not in the document]
Comment (Ayan-10) [reply app/src/main/res/layout/activity_custom_selector.xml:37]: I didn't get the 5dp part. Can you please elaborate?

File: app/src/main/java/fr/free/nrw/commons/customselector/database/NotForUploadStatus.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package fr.free.nrw.commons.customselector.database
+
+import androidx.room.*
+
+/**
+ * Entity class for Not For Upload status.
+ */
+@Entity(tableName = "not_for_upload_table")
```
[the remark this answers is not in the document]
Comment (nicolas-raoul) [reply app/src/main/java/fr/free/nrw/commons/customselector/database/NotForUploadStatus.kt:8]: `not_for_upload_table` is consistent with `depicts_table` and `uploaded_table`.
`NotForUploadStatus.kt` is consistent with `UploadedStatus.kt`.
So it looks good to me.

File: app/src/main/res/layout/activity_custom_selector.xml (hunk 1 above)
Comment (nicolas-raoul) [reply app/src/main/res/layout/activity_custom_selector.xml:20]: Sounds like a good idea!

File: app/src/main/java/fr/free/nrw/commons/customselector/database/NotForUploadStatus.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (4D17Y4) [reply app/src/main/java/fr/free/nrw/commons/customselector/database/NotForUploadStatus.kt:8]: Nice.

File: app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/CustomSelectorActivity.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (4D17Y4) [reply app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/CustomSelectorActivity.kt:131]: Yes, thanks

File: app/src/main/res/layout/activity_custom_selector.xml [hunk 5]
```diff
@@ -14,6 +14,52 @@
...
+      android:visibility="gone"
+      app:layout_constraintBottom_toBottomOf="parent"
+      app:layout_constraintEnd_toEndOf="parent"
+      app:layout_constraintStart_toStartOf="parent">
+
+        <Button
+          android:id="@+id/not_for_upload"
+          android:layout_width="0dp"
+          android:layout_height="wrap_content"
+          android:text="Mark as not for upload"
```
Comment (4D17Y4) [inline app/src/main/res/layout/activity_custom_selector.xml:34]: Similarly for the text, you can use @string/<name of your variable>, 
This is very important so as to make the application multilingual. If you use English here, the text will always be in English. But if you declare them as string constants, the value would be translated by the translated wiki to different languages. Thus when a user changes the app language, all the text in the application would be derived from the specific language strings.

File: app/src/main/res/layout/activity_custom_selector.xml [hunk 6]
```diff
@@ -14,6 +14,52 @@
...
+          android:layout_width="wrap_content"
+          android:layout_height="wrap_content"
+          android:orientation="vertical"
+          app:layout_constraintGuide_percent="0.5" />
+
+        <Button
+          android:id="@+id/upload"
+          android:layout_width="0dp"
+          android:layout_height="0dp"
+          android:text="Upload"
```
Comment (4D17Y4) [inline app/src/main/res/layout/activity_custom_selector.xml:54]: Same as the previous one.

**Review by 4D17Y4 — CHANGES_REQUESTED** [review]
Some additional changes.

File: app/src/main/res/layout/activity_custom_selector.xml (hunk 5 above)
Comment (nicolas-raoul) [reply app/src/main/res/layout/activity_custom_selector.xml:34]: As we are in an early phase, doing quick prototyping without proper internationalization is OK as long as you don't forget it later, especially for things that might be removed/modified later.
I believe this text has high chances of making it into master, though :-)

File: app/src/main/res/layout/activity_custom_selector.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ayan-10) [reply app/src/main/res/layout/activity_custom_selector.xml:37]: Oh okay. Got it.

**Review by 4D17Y4 — APPROVED** [review]

**Comment by nicolas-raoul** [comment]
Minor remark: To avoid technical problem on various operating systems, we try to avoid uppercase and especially apostrophes in branch names ([examples](https://github.com/commons-app/apps-android-commons/branches)).
Apostrophes have a tendency to trigger many bugs 😅

**Comment by nicolas-raoul** [comment]
Merged into https://github.com/commons-app/apps-android-commons/tree/gsoc22

**Pull request closed without merging**